src/util/validate.h

```diff
@@ -0,0 +1,99 @@
+/*******************************************************************\
+
+Module: Goto program validation
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_VALIDATE_H
+#define CPROVER_UTIL_VALIDATE_H
+
+#include <type_traits>
+
+#include "exception_utils.h"
+#include "invariant.h"
+#include "irep.h"
+
+enum class validation_modet
+{
+  INVARIANT,
+  EXCEPTION
+};
+
+#define GET_FIRST(A, ...) A
+
+/// This macro takes a condition which denotes a well-formedness criterion on
+/// goto programs, expressions, instructions, etc. Based on the value of the
+/// variable vm (the validation mode), it either uses DATA_INVARIANT() to check
```

Comment:
I may have missed a global declaration of `vm` somewhere, but from what I've seen it looks like `vm` is just 'whatever variable of that name is in scope at the moment?` - Could you either make that explicit in the docs and/or actually add a global declaration for it somewhere? I'm otherwise wondering what the compile error messages might look like if someone tries to use these macros outside of a validate() function, or if the try to declare a variable of a different type but the same name... :-)

Reply:
Ok, I've added some doc and added a static assertion to check that `vm` has the expected type.